typescript/infra/scripts/helloworld/kathy.ts

```diff
@@ -54,51 +54,79 @@ async function main() {
     throw new Error(`Invalid chains to skip ${invalidChains}`);
   }
 
-  let failureOccurred = false;
-
   const origins = chains.filter((chain) => !skip || !skip.includes(chain));
+  const pairings = diagonalize(
+    origins.map((origin) =>
+      origins.map((destination) =>
+        origin == destination ? null : { origin, destination },
+      ),
+    ),
+  )
+    .filter((v) => v !== null)
+    .map((v) => v!);
+
+  // default to once every 6 hours getting through all pairs
+  const fullCycleTime = process.env['KATHY_FULL_CYCLE_TIME']
+    ? parseInt(process.env['KATHY_FULL_CYCLE_TIME'])
+    : DEFAULT_FULL_CYCLE_TIME;
+  if (!Number.isSafeInteger(fullCycleTime) || fullCycleTime <= 0) {
+    error('Invalid cycle time provided');
+    process.exit(1);
+  }
 
-  // submit frequently so we don't have to wait a super long time for info to get into the metrics
-  const metricsInterval = setInterval(() => {
-    submitMetrics(metricsRegister, 'kathy', { appendMode: true }).catch(
-      console.error,
-    );
-  }, 1000 * 30);
-
-  for (const origin of origins) {
-    for (const destination of origins.filter((d) => d !== origin)) {
-      const labels = {
+  // track how many we are still allowed to send in case some messages send slower than expected.
+  let allowedToSend = 0;
+  setInterval(() => {
+    allowedToSend++;
+  }, fullCycleTime / pairings.length);
+
+  for (
+    // in case we are restarting kathy, keep it from always running the exact same messages first
+    let currentPairingIndex = Date.now() % pairings.length;
+    ;
+    currentPairingIndex = (currentPairingIndex + 1) % pairings.length
+  ) {
+    currentPairingIndexGauge.set(currentPairingIndex);
+    // wait until we are allowed to send the message
+    while (allowedToSend <= 0) await sleep(1000);
+    allowedToSend--;
+
+    const { origin, destination } = pairings[currentPairingIndex];
+    const labels = {
+      origin,
+      remote: destination,
+    };
+    try {
+      await sendMessage(app, origin, destination, gasCalc);
+      log('Message sent successfully', { origin, destination });
+      messagesSendCount.labels({ ...labels, status: 'success' }).inc();
+    } catch (e) {
+      error(`Error sending message, continuing...`, {
+        error: e,
```

Comment:
I learned this will log the error as `'{}'`, see https://github.com/abacus-network/abacus-monorepo/pull/843/files#diff-10f7a1d27ccc4927804c7965831ff7b41986237dab3a841be1378911733293f3R251 for the fix

Between this and the new log file in here, I feel like we should more seriously consider just moving to a lightweight library like Bunyan in the future. But for now I'm fine with the logging in here